feat: settlement contract wrapper (#1)
cowdao-grants/rfps: RFP-01.md

- $50k in yield-bearing stable coins or ETH
- 500k COW

The target deposit is:

- $100k in yield-bearing stable coins or ETH
- 1M COW

## Goal

The goal of this project is to implement a `SubPoolFactory` that allows solvers to participate in the batch auction with reduced bonding requirements. The `SubPoolFactory` will deploy a `SubPool` for each solver, which will be funded to at least the minimum required amount. The `SubPoolFactory` will also provide a wrapper around the `GPv2Settlement` contract that allows solvers to execute their solutions by providing a signature attesting to the solution and their parameters.

### Key Features

- Ultra gas efficient signed settlement function
- No co-mingling of funds between solvers
- Easily observable on-chain addresses for MEV Blocker rebates
- Permissionless access to the batch auction with reduced bonding requirements

### Deliverables

- Contract system for `SubPoolFactory`, `SubPool`, and `SignedSettlement`
- Comprehensive documentation
- Open-source code repository

### Interface

```solidity
interface Auth {
    /**
     * Allows `usr` to perform root-level functions
     */

>     function rely(address usr) external auth;

Reviewer: ```suggestion
    function addOwner(address usr) external auth;
```

    /**
     * Denies `usr` from being able to perform root-level functions
     */

>     function deny(address usr) external auth;

Reviewer: ```suggestion
    function removeOwner(address usr) external auth;
```

    /**
     * Determine if `usr` is able to perform root-level functions
     */

>     function wards(address usr) external view returns (bool);

Reviewer: ```suggestion
    function isOwner(address usr) external view returns (bool);
```

It could all be so simple for contributors with a B1 level in english.

}

> interface SubPool is Auth {

Reviewer: Why does this need to be auth (looks like there is nothing that requires only owner).

I guess we want to allow changes to the backend url?

Reviewer 2: Will need it for announceDestroy too. But we can move all the auth calls to the factory with a limitation that only the owner would be able to call those. With Auth implemented the pool owner can allow other addresses to do those actions.

Author: The reason for `auth` is that the way `auth` is designed allows for _multiple_ (what you call) owners. I think root level functionality on sub-pools should _not_ be bound to the specific EOA that is `tx.origin` on the transaction.

Consider the case of a solver who is using a hardware signing key, or a service like AWS KMS for signing. Having to then use those keys to adjust the backend URI, or exit may be onerous. So this method would allow for a subpool to be bound to a specific signing key for submitting the settlement, however also allow for other ways for them to administer their sub pool through whatever method they want.

If this is the case then, having the auth list inside the SubPool is a little nicer IMO than having a multi-dimensional auth mapping in `SubPoolFactory` and allows for a "separation of concerns" with respect to authorisation (if you want to know if address `x` is authorised on the sub pool - you ask the sub pool).

Reviewer: I see, but we also wanted a function to determine for a given address what is the pool it belongs to, right (so that it's easy for someone handing out offchain signatures to a solver EOA to check if they are still collateralized)?

Are we doing this with a map inside the factory?

Author: > Are we doing this with a map inside the factory?

I'd leave this open to the grantee. This could be implemented using either deterministic addresses, or a mapping. Functions on the factory are `solverSubPool` (to get from solver EOA -> pool address), and `canSolve` which does the view on-chain validation that the solver is collateralised at the level required (and has a valid sub pool etc).

Author: Collateralisation and types, as the sub-pools are KYC'ed is now delegated to off-chain handling.

    /**
     * Determine the amount required to `heal` the pool
     * @return amt of immutable collateralToken
     * @return cowAmt of COW
     */
    function dues() external view returns (uint256 amt, uint256 cowAmt);
    /**
     * Pull in from `msg.sender` the dues required to `heal` the pool
     */
    function heal() external;
}

interface SubPoolFactory is Auth {
    // --- view functions ---

    /**
     * View function for determining the minimum collateral required if using `token`
     * @param token The token to retrieve the collateral requirement for
     */

>     function collateral(IERC20 token) external view returns (uint256);

Reviewer: This should probably list all collateral requirements with a single call for convenience.

Author: Is the requirement for arrays or linked lists really worth it?

Reviewer: Maybe not, but now you require extra indexing logic to find out what are the required collaterals, whereas for relatively little gas you could just expose it on chain. No strong feelings.

    /**
     * View function for determining if a given solver EOA has a correctly collateralized
     * sub-pool and has not announced an exit.
     * @param subPoolSolver The solver to check
     * @return bool if the solver is allowed to solve
     */

>     function canSolve(address subPoolSolver) external view returns (bool);

Reviewer: Should we also provide a view function to get the pool address for a specific subPoolSolver?

Author: Sure!

Author: Added, please review and mark as resolved if no issues :pray:

     * @return string URI
     */
    function backendUri(SubPool pool) external view returns (string memory);

    // --- permissionless functions ---
    /**
     * Deploy a `SubPool` deterministically based on `msg.sender`.
     * @notice The `msg.sender` SHOULD NOT be a contract.
     * @param token The nominated token to use for collateral
     * @param amt How much collateral to add to the sub-pool (must be minimally collateralised)
     * @param cowAmt Amount of COW to add as collateral
     * @param backendUri The URI to send the batches to the solver
     * @dev Emits an event `SubPoolCreated` when creating a sub-pool
     */
    function create(IERC20 token, uint256 amt, uint256 cowAmt, string calldata backendUri) external;

    // --- cow dao bonding pool authed functions ---

    /**
     * Set the delay for a sub-pool to exit
     * @dev Emits an event when setting a delay
     * @param delay The delay in seconds
     */
    function setExitDelay(uint256 delay) external auth;
    /**
     * Allow `token` to be used as collateral for the sub-pool.
     * @dev Emits an event `AllowCollateral` when allowing/updating a token
     */

>     function allowCollateral(IERC20 token, uint256 min) external auth;

Reviewer: If there are multiple collaterals do all need to be provided or some (maybe document)?

Author: > If there are multiple collaterals do all need to be provided or some (maybe document)?

By multiple collaterals, do you refer to:
i. The a subpool factory **owner** (such as CoW DAO bonding pool) allowing subpools with different collaterals, but where each subpool itself consists only of collateral 'A' + COW; or
ii .The subpool itself may be collateralised with tokens X + Y + Z + COW?

Reviewer 3: I agree this is very unclear.
We want to ask a solver for COW plus USDC at the same time as collateral. In particular, `create` is going to take two tokens, one (likely COW) is implicit.
I imagine if I call `allowCollateral(MILK, 1337)` then `create` can be called for either `token == USDC` or `token == MILK`? This is how I interpret this function, in any case it should be documented more clearly.

Note also that, if what I surmised above is right, there's no function that allows the DAO to change the minimum amount of collateral. Is the minimum amount of COW immutable and decided on deployment?

Author: I think more importantly, we need to have some idea of how to handle the ETH from the MEV Blocker rebates (this adds a third collateral potentially). @fleupold how did you envision that this would work?

Also, particularly as well, what should happen in the case that the solver is building from the small bond to the target bond, and they choose to withdraw prior to the 365 day period ending?

Reviewer 2: This does seem important. How important is the 365 day period? And what do we want in the contracts related to that? For handling eth, seems like there are two options here, at exit solver gets their collateral token, cow token and eth back or they just get the collateral and cow tokens back without the eth. In that case we should also have some config regarding where the eth should be sent to?

Reviewer: > How important is the 365 day period?

I would not enshrine the 365 day period in the contracts anywhere (I think the signer can simply seize to provide signatures to solvers that have not reached their target deposit after a certain time). If a solver wants to withdraw they should be able to withdraw any token balance they hold after the challenge period.

As to how to deal with the MEV Blocker ETH kickbacks, I see two options:
1. Add ETH as a collateral type that can be used instead of yield bearing USD
2. Have the refund recipient be a smart order that automatically wraps and swaps ETH for yield bearing USD with `receiver` being the solvers pool (using e.g. a `TradeAboveThreshold` smart order with pre-interactions)

Maybe we should even get rid of `collateral`, `allowCollateral`, etc. and have the collateral requirements be communicated off-chain as well? Otherwise we would definitely need a more sophisticated structure here (e.g. you need this much CoW and one of the following (50k USDC, USDT or 50 ETH).

Author: > I would not enshrine the 365 day period in the contracts anywhere (I think the signer can simply seize to provide signatures to solvers that have not reached their target deposit after a certain time). If a solver wants to withdraw they should be able to withdraw any token balance they hold after the challenge period.

This lacks certainty as to what happens to funds in the case of:

1. Withdrawal prior to the end of the 365 day period - what happens to accumulated funds over the minimum?
2. What happens after the 365 day period and the solver has failed to reach the target?

The 365 days is enshrined in CIP-44, and the above questions are very reasonable for a prospective solver to ask before even committing funds to start with. 

I do fear that by moving more and more elements off-chain will present an onerous administrative burden and that we should seize this initiative for true permissionless (save bond) access to the batch auction.

Reviewer: Sorry for not being clear

> 1. Withdrawal prior to the end of the 365 day period - what happens to accumulated funds over the minimum?

They go back to the solver

> What happens after the 365 day period and the solver has failed to reach the target?

Whatever is providing signatures for the solver to use the settlement wrapper contract, seizes to do so leaving the solver with the choice to leave their capital idle in the pool or withdraw eventually.

--- 
I fear that by putting all the logic we know of right now on chain (which inevitably will be a subset of the logic the bonding pool will enforce in 6 months as requirements change) we will end up with a hybrid system where whatever off-chain component is handing out signatures is doing some check off-chain and some checks on-chain (and it won't necessarily be clear to the solver why they are not receiving signatures from on-chain state only and even if there is no guarantee the signer cannot start withholding signatures for a solver it no longer likes). For complete permissionless-ness, solvers need to create a full bonding pool.

However, since signature under-writing has to happen off-chain I think it makes sense to say "the component that underwrites has certain criteria that it publishes somewhere publicly" and that's the source of truth. IMO it's just a waste of time and complexity to enshrine things like exact collateral requirements, timelines on when those should be reached, etc. into this wrapper contract.

Imo the main goal of the Subpool contract is to replace the Safes we currently use to store collateral with an onchain system that has better custody (ie a way to somewhat permissionless-ly withdraw) and allows solvers to adjust frequently changing aspects about their pool (e.g. submission address, backend url) without intervention from our side. But really the more important contribution here is the wrapper contract...

Author: OK, now this follows a `bill` and `dues` methodology and `canSolve` will be `false` if there is an outstanding `bill`, or the solver has announced and exit. This allows for arbitrary collateralisation (with the mandate of a minimum amount of `cow` for spam prevention). The rest is off-chain.

     * @param amts The amount of tokens to fine
     * @dev Emits an event `Fine` when fining a sub-pool
     */
    function fine(SubPool pool, IERC20[] calldata tokens, uint256[] calldata amts) external auth;

    // --- SubPool authed functions ---

    /**
     * Indicate a sub-pool's intent to quit the batch auction
     * At this point the solver whose sub-pool it is is now removed from
     * the competition.
     * This then starts the `delay` timer until exit is allowed.
     * @param pool The sub-pool to announce an exit for
     * @dev Emits an event `AnnounceExit`
     */
    function announceDestroy(SubPool pool) external onlySubpoolAuth;
    /**
     * Withdraw any remaining collateral / cow from the sub-pool.
     * Funds are sent to `msg.sender`.
     * @dev Should emit an event `SubPoolDestroyed` on the factory
     */
    function finalizeExit(SubPool pool) external onlySubpoolAuth;
    /**
     * Set the backend URI for the sub-pool
     * @param pool The sub-pool to set the URI for
     * @param newUri The new URI to set
     * @dev Emits an event `SetBackendUri` on the factory when setting a URI
     */

>     function setBackendUri(SubPool pool, string calldata newUri) external onlySubpoolAuth;

Reviewer: nit: maybe this doesn't need auth if it implies you can only update your own pool's uri anyway (could compute pool from message.sender)

Author: See previous comment on multiple owners

}

interface SignedSettlement is Auth {
    /**
     * Signed settlement function that is callable by `msg.sender` if `msg.sender` is in
     * receipt of a valid signature. The signer is a trusted entity and bears the responsibility
     * of ensuring that `msg.sender` is correctly collateralised.
     *  
     * The signed data **MUST** include:
     * - `solver` - the address of the sub-bonded solver, i.e. `msg.sender`

>      * - Replay protection - an expiry based on some configurable method (e.g. block number)

Reviewer 2: I think replay protection is built in because you wouldn't be able to execute same orders again. We can add an expiry on the validity of the solution and signature though.

Author: This is not necessarily guaranteed as the orders could be partially fillable (and not yet completely filled).

Reviewer 2: Ah, okay, we can add a nonce to the message being signed, but that is going to need an SSTORE post successful execution and an SLOAD pre-execution. An expiry won't necessarily protect against replay, because it can be replayed before the expiry. I think if we want 0 SSTORE/SLOAD this replay violation needs to be dealt offchain with fines probably?

Author: > I think if we want 0 SSTORE/SLOAD this replay violation needs to be dealt offchain with fines probably?

This is kind of my thinking where we should have some relatively good guarantees (like the signature is only good for 'x' blocks which gives them an opportunity to land it on chain). 

If it lands on-chain multiple times, then it would be an off-chain fine / slashing event. Otherwise, we hit the SSTORE / SLOAD issue.

     *
     * @notice This contract is a facade around the `GPv2Settlement` contract and has no
     *         on-chain state / binding to the concept of a sub-pool. It is merely a
     *         permissioned wrapper around the `settle` function.
     * @dev There must be **NO** use of `SLOAD` / `SSTORE` in the execution path.
     */
    function signedSettle(
        bytes calldata signedData,
        bytes calldata unsignedData,

>     ) external onlySubPool;

Reviewer 2: Need to decide what parameters are part of the signature. Still not clear if we are including interactions and which interactions?

Reviewer 2: @fleupold

Reviewer 3: This decision is key and should be included in the document. This is what I think should be included in the signature:

```solidity
struct SignatureParts {
  /// The solver address that should be checked against `msg.sender``.
  address solver;
  /// The signature is invalid after this time.
  /// Note that signatures can be replayed until the deadline passes.
  uint32 deadline;
  // Parameters from the call to `settle`
  IERC20[] calldata tokens;
  uint256[] calldata clearingPrices;
  GPv2Trade.Data[] calldata trades;
}
```

This should not be, though it will be included in the function input in some form:
```solidity
struct NotSigned {
  // Parameters from the call to `settle`
  GPv2Interaction.Data[][3] calldata interactions;
}
```

I wrote them as Solidity struct for clarity but they don't have to be, I expect them to be packed in some reasonable format.

Reviewer 3: I also don't think we should enforce a function signature for `signedSettle` but have it depend on the gas analysis done during implementation for optimization purposes.

Thinking about a concrete implementation, a reasonable function format could be something like:
```
function signedSettle(
  bytes memory data
) external;
```

Notably, `onlySubPool` is missing since the check is done though the signature and `SignedSettlement` doesn't know about any subpool.

`data` could be `abi.encodePacked(signature, deadline, abi.encodeCall(ISettlement.settle, (...settlement input)))`. Then some minimal memory manipulation would allow to hash content equivalent to `SignatureParts` before doing a low-level call with the data already in memory.
Then, by how `SignatureParts` is defined, we can just (1) change the memory allocated by Solidity before `deadline` in `data` to include the eth-sign message plus `msg.sender`, and (2) hash the data until the memory start of `interactions` (the length of each field but `interactions` would be hashed as well since the interaction offset is in there and hashed as well). If my understanding of ABI encoding is right this should be a valid eth-call signature that requires minimal memory changes before forwarding the call to `settle`.

Considering how narrow the scope of this contract is, we could even write it directly in yul and skip Solidity.

Reviewer 2: I don't think you will be able to hash in-place for 712 because each element of the struct will need to be hashed and then hash the hashes, etc? But yes, we should reuse the abi encoding instead of encoding the args again since its the same structures and order of args, just need to patch a selector at the start.

Reviewer 2: Hmm, i just noticed you mentioned eth-sign so assuming 191 we can do minimal memory manipulation to get the digest to sign indeed 👍.

Author: I don't think we should be going into _concrete_ implementation details here (i.e. yul suggestions etc). This is for a grant afterall (and they're paid for this :laughing:)

Having said this - what needs to be signed is important. I agree with the `SignatureParts` struct suggestion, however @fleupold made mention of the interactions - notably there should be some binding restriction that a user's pre and post interactions are included (so CoW DAO bonding pool meets it's social consensus obligations). So this muddies it a bit as to do:

1. Just pre/post; or
2. All interactions

The outcome of (1) or (2) above I think would likely rest in gas profiling, so I think being overly specific as to how the implementation should work isn't what we should do here and just kick it back to the grantee.

Reviewer 3: Yeah, the only part that should appear in the document is details on what has to be signed, I got carried away. :slightly_smiling_face: 
A takeaway from this was that for something that needs to be gas efficient, we should probably not specify an interface at all but just say what we need. Then the actual implementation can be optimized without having to fit an interface. It's a different thing for the other contracts where gas efficiency is less of a concern.

Reviewer 2: if we using eth-sign aka 191 including interactions in the signature is trivial, for 721 you need to do more stuff and more hashes gas wise. Of course with number of interactions and size of interactions the gas for hash will scale linearly too.

Reviewer 3: The main reason why we wanted to avoid signing interactions in the first place is gas efficiency, less things to hash. On the other hand, the original concern was that we'd have to write to memory twice, whose cost grows quadratically with size. Writing once is unavoidable because of the call, but if we don't need to write to memory more then we only need to work with the cost of extra hashing, 6 bytes per word.

In the last 20 transactions, I see that this would be more or less between 207 and 669 extra gas, which I think is acceptable. So maybe we should hash everything.

<details><summary>analysis of random transactions</summary>

["Lean" transaction](https://etherscan.io/tx/0x649295f62bac4c971c5564830c44cff32d66dde77772adc383317bd3e2af706c) (just Uniswapv2)
Calldata bytes: 2092
Bytes dedicated to everything but interactions: 0x3e0 = 992
Cost of extra hashing: 6 · ceil((2092 - 992)/32) = **207**

["Fattest" transaction](https://etherscan.io/tx/0xd7f21343c59f1b215e84d0c0b29462ad29779b4c364f7808c04ceb0488991a90) among the most recent 20.
Calldata bytes: 4556
Bytes dedicated to everything but interactions: 0x3e0 = 992, again
Cost of extra hashing: 6 · ceil((4556 - 992)/32) = **669**

</details>

Reviewer 4: >   /// The signature is invalid after this time.
  /// Note that signatures can be replayed until the deadline passes.
  uint32 deadline;
  
 Is there a specific reason to make this a timestamp instead of a block? If it were a block there would be less uncertainty whether it still makes sense for the solver to submit the solution. The less predictable the block times of a chain are the less clear the submission strategy becomes IMO.
 Additionally this would already fit nicely into a possible 1 block per auction world.
 
Why should interactions not be part of the commitment? I think one of the warts of the current contract is that there are no guarantees around pre- and post-interactions. Verifying these would resolve this, no?

Reviewer: If possible, we shouldn't require signing all interactions. The `driver` doesn't report their whole solution back to the autopilot, therefore the autopilot cannot sign off on all pre/post interactions a solver may want to include (e.g. for just-in-time CoW AMM orders). While we could change this, this would require a larger update to the existing driver<>autopilot communication.

What the autopilot does know is which user orders are going to be settled and whether those have `pre` or `post` interactions (available in the orderbook DB).

Ideally it would commit to just this subset of interactions, so that the wrapper contract can enforce inclusion of user specified pre and post interactions.

Author: > If possible, we shouldn't require signing all interactions. The driver doesn't report their whole solution back to the autopilot, therefore the autopilot cannot sign off on all pre/post interactions a solver may want to include (e.g. for just-in-time CoW AMM orders). While we could change this, this would require a larger update to the existing driver<>autopilot communication.

For this, I think we should be governed by gas considerations here. Let's make it part of the spec that the grantee should assess both methods for gas optimisation, and then make a decision on the analysis.

### Key Features

- Ultra gas efficient signed settlement function
- No co-mingling of funds between solvers
- Easily observable on-chain addresses for MEV Blocker rebates
- Permissionless access to the batch auction with reduced bonding requirements

### Deliverables

- Contract system for `SubPoolFactory`, `SubPool`, and `SignedSettlement`
- Comprehensive documentation
- Open-source code repository

### Interface

```solidity
interface Auth {
    /**
     * Allows `usr` to perform root-level functions
     */
    function addOwner(address usr) external auth;
    /**
     * Denies `usr` from being able to perform root-level functions
     */
    function removeOwner(address usr) external auth;
    /**
     * Determine if `usr` is able to perform root-level functions
     */
    function isOwner(address usr) external view returns (bool);
}

> interface SubPool is Auth {

Reviewer 3: Why does a subpool need to exist? Can't everything be done with internal accounting in the factory?

Author: This is done so that MEV blocker kickbacks can be sent to the address which simplifies accounting instead of this being all co-mingled. This is also somewhat of an oversight (ETH handling) and something that I'd prefer to kick back to the grantee to work out how to handle the additional collateral.

Reviewer 3: In principle this can be handled as well with internal balances (`deposit(target[], amount[])`). But if it's easier to integrate this with other tools, I'm ok with keeping subpool. We should still add the ETH handling (and arbitrary ERC20 as well?) to the scope.

Author: > In principle this can be handled as well with internal balances (deposit(target[], amount[])). But if it's easier to integrate this with other tools, I'm ok with keeping subpool. We should still add the ETH handling (and arbitrary ERC20 as well?) to the scope.

In principle, yes, but this requires more integration, no? If having a subpool address, the solver can simply specify the sub pool address in the MEV blocker RPC URL and this will cause the rebate to be sent there (just a simple transfer). Then no co-mingling of funds, or requirement for craz(ier) accounting?

    /**
     * Determine the amount required to `heal` the pool
     * @return amt of immutable collateralToken
     * @return cowAmt of COW
     */
    function dues() external view returns (uint256 amt, uint256 cowAmt);
    /**
     * Pull in from `msg.sender` the dues required to `heal` the pool
     */
    function heal() external;
}

interface SubPoolFactory is Auth {
    // --- view functions ---

    /**

>      * View function for determining the minimum collateral required if using `token`

Reviewer 3: I'd make what happens in the bad case explicit.

```suggestion
     * View function for determining the minimum collateral required if using `token`
     * If the token is not a valid collateral, this function reverts.
```

     * @param token The token to retrieve the collateral requirement for
     */
    function collateral(IERC20 token) external view returns (uint256);
    /**
     * View function for determining if a given solver EOA has a correctly collateralized
     * sub-pool and has not announced an exit.
     * @param subPoolSolver The solver to check
     * @return bool if the solver is allowed to solve
     */

>     function canSolve(address subPoolSolver) external view returns (bool);

Reviewer 3: Many nits: not mentioning EOA anymore (I see we want this, but there's no need to enforce it here since this isn't something that can happen based on `create` checks), nor "view" (redundant), also not assuming that the input is a solver and explaining what a "solver" actually is.

```suggestion
    /**
     * Determine if an address has a collateralized sub-pool and has not announced an exit.
     * While not enforced on-chain, an address for which this call returns `true` is a
     * solver and is allowed to submit settlement on CoW Protocol through
     * `SignedSettlement`.
     * @param candidate The address to check
     * @return bool if the address should be allowed to submit settlements through
     * `SignedSettlement`
     */
    function canSolve(address candidate) external view returns (bool);
```

    /**

>      * View function for determining the sub-pool for a given solver

Reviewer 3: I assume this is the intended behavior and would like to see this made explicit.

```suggestion
     * View function for determining the sub-pool for a given solver.
     * Reverts if the input address has no sub-pool to its name.
```

     * @param solver The solver to retrieve the sub-pool for

>      * @return SubPool

Reviewer 3: ```suggestion
     * @return The contract managing the collateral paid by this solver to be
     * allowed to settle transactions on CoW Protocol
```

     */
    function solverSubPool(address solver) external view returns (SubPool);
    /**
     * View function for determining the backend URI for the sub-pool
     * @return string URI
     */

>     function backendUri(SubPool pool) external view returns (string memory);

Reviewer 3: Do we want to make some affordances for solvers who wish to keep this secret? We could let them encrypt their URI with the autopilot's private key.

Author: No, I think signing / secrets should not be the domain of this contract. For permissioned access this should be something like SSL certificates (PKI) or similar. This is not in scope IMO.

    // --- permissionless functions ---
    /**
     * Deploy a `SubPool` deterministically based on `msg.sender`.

>      * @notice The `msg.sender` SHOULD NOT be a contract.

Reviewer 3: Should not or must not?
Do we mind if solvers are contracts? Unless there are unknown vulnerabilities in the settlement contract, it should be ok to have solver contracts.
I admit it makes it a bit easier to exploit vulnerabilities somewhere more easy, but not too much since most useful call can be made inside the settlement contract as well.
If motivated, I'd make it a MUST. Otherwise I'd consider allowing contracts.

Author: This specification is written for someone executing on a grant. "SHOULD" means they have leeway (this isn't a protocol specification - it's a grant - this isn't an interface of a multilateral agreement among many, many stakeholders).

So, IMO **if** the implementer wants to include contracts, so be it, but it would be on them to ensure that vulnerabilities are assessed / taken care of.

Reviewer 3: To me this is a protocol-level decision: a grantee has no way to judge if in the full scope of the protocol having a smart-contract solver could be an issue. Also, most concerns on the protocol side are imho for ancillary tools like analytics that might be assuming something about the transaction sender, or that there can be at most a settlement per transaction.
Pragmatically, a grantee isn't going to take the time to weight these considerations. So it's on us: do we want it or not?

Author: I don't think this is a _protocol-level_ decision but is a _bonding pool_ level decision. By virtue of having the `SignedSettlement` contract, running under CoW DAO's bonding pool, the protocol-level decision is implicitly made.

Having said this, I know that the multiple settlements per transaction have historically caused issues and this is an easy one to overlook, so I'd be inclined towards making this a **MUST NOT** be a contract?

     * @param token The nominated token to use for collateral

>      * @param amt How much collateral to add to the sub-pool (must be minimally collateralised)

Reviewer 3: What about not specifying this and just assuming that the input is `collateral(token)`?
If this is needed anyway, I suggest the following:
```suggestion
     * @param amt How much collateral to add to the sub-pool (the call will revert if
     * this value is below the minimal collateral needed for `token`)
```

     * @param cowAmt Amount of COW to add as collateral
     * @param backendUri The URI to send the batches to the solver
     * @dev Emits an event `SubPoolCreated` when creating a sub-pool
     */
    function create(IERC20 token, uint256 amt, uint256 cowAmt, string calldata backendUri) external;

    // --- cow dao bonding pool authed functions ---

    /**

>      * Set the delay for a sub-pool to exit

Reviewer 3: ```suggestion
     * Set the minimum delay before a solver can claim back the collateral from its
     * sub-pool
```

     * @dev Emits an event when setting a delay
     * @param delay The delay in seconds
     */
    function setExitDelay(uint256 delay) external auth;
    /**
     * Allow `token` to be used as collateral for the sub-pool.
     * @dev Emits an event `AllowCollateral` when allowing/updating a token
     */
    function allowCollateral(IERC20 token, uint256 min) external auth;
    /**
     * Revoke `token` as a collateral for sub-pools
     * @dev Emits an event `RevokeCollateral` when revoking a token
     */
    function revokeCollateral(IERC20 token) external auth;
    /**
     * Fine a sub-pool
     * @param pool Which sub-pool to fine
     * @param tokens The tokens to fine (sending to `msg.sender`)
     * @param amts The amount of tokens to fine

>      * @dev Emits an event `Fine` when fining a sub-pool

Reviewer 3: ```suggestion
     * Fine a sub-pool and send the proceeds to the caller
     * @param pool Which sub-pool to fine
     * @param tokens The tokens to fine
     * @param amts The amount of tokens to fine
     * @dev Emits an event `Fine` when fining a sub-pool
```

     */

>     function fine(SubPool pool, IERC20[] calldata tokens, uint256[] calldata amts) external auth;

Reviewer 3: I suggest to also include a "freeze" and "unfreeze" function that disallow the solver to exit but still don't cause funds to be taken away from the pool. This is done for two reasons:
1. Assumption of innocence: give time to start an investigation without having to take the money in advance.
2. Security: if the collateral token cannot be transferred (say, because USDC froze any onchain transfer for whatever reason) then this function would not be callable and once transfers resume then the exit time might have passed and a malicious solver could exit before the protocol fines it.

Author: > Assumption of innocence: give time to start an investigation without having to take the money in advance.

The same can be done at an off-chain level with just refusing to sign submissions from the sub-pool bonded solver.

> Security: if the collateral token cannot be transferred (say, because USDC froze any onchain transfer for whatever reason) then this function would not be callable and once transfers resume then the exit time might have passed and a malicious solver could exit before the protocol fines it.

This also breaks quite important security assumptions from the _solver's_ perspective. What's to say CoW DAO couldn't just freeze their funds indefinitely with no end in sight? On net, I think it's fair to _not_ include a `freeze` / `unfreeze` as (1) is void and (2) structures the deal onerously in favour of CoW DAO's bonding pool (the delay should be set at such a level as to provide a reasonable guarantee to CoW DAO that there is time to analyze the behaviour and fine them if necessary, but ALSO be respected from the solver's perspective that if they want to leave - they have a guaranteed time at which they am assured of receiving collateral).

Reviewer 3: In short: if the exit time is long, solvers won't like to put their money in the contract. If it's short, the DAO doesn't have time to investigate and needs to take the money first and think later (the time for a proposal is already a week). I believe solvers would be happier with a shorter time but the possibility of an undetermined freeze. Anyway, we could ask them.

> This also breaks quite important security assumptions from the solver's perspective.

A hostile DAO can just take the money, so the security model for a solver already requires the DAO to be largely trusted. Freezing/unfreezing doesn't change the attack surface for the solvers except for more smart-contract risk.

> (1) is void and (2) structures the deal onerously in favour of CoW DAO's bonding pool

For (2), the deal is already onerous in favor of the bonding pool (better, the DAO? I assumed the fining was from the DAO but maybe I'm mistaken; all points still apply anyway). Not sure about the voidness, but my "attack" seems concrete enough to be a worry: a solver sees some issues with USDC, exploits something that it knew was possible but wasn't worth the bond money, and then exits, trying to take the money before us once USDC is tradable again.

> the delay should be set at such a level as to provide a reasonable guarantee to CoW DAO that there is time to analyze the behavior and fine them if necessary, but ALSO be respected from the solver's perspective that if they want to leave 

Seen these concerns, we should add a concrete suggestion for its value in this proposal. From the look of it, considering the time needed for a DAO to act, a few months should be good?

Having a freezing/unfreezing allows for a much shorter exit time (two weeks?). If a solver tries to leave when something suspicious is happening, funds can be freezed and the investigation can be resumed.

If we're concerned about long investigation times (which is fair from a social perspective) we could add a longer timeout after which the funds are released if no action is taken. It seems unlikely to me but I admit it's hard to judge.

Reviewer 2: I think a few months time from announcing exit to withdrawing might be long. Ideally, there is system in place that triggers some automated checks on the historical settlements of the particular solver or prompts some dao committee to verify manually if the solver is safe to leave the system. Longer timeouts might also lead to procrastination.

Author: > For (2), the deal is already onerous in favor of the bonding pool (better, the DAO? I assumed the fining was from the DAO but maybe I'm mistaken; all points still apply anyway). Not sure about the voidness, but my "attack" seems concrete enough to be a worry: a solver sees some issues with USDC, exploits something that it knew was possible but wasn't worth the bond money, and then exits, trying to take the money before us once USDC is tradable again.

The USDC vector is similiar in level of a chosen collateral itself going to zero, in which case the risk falls down to the bonding pool (CoW DAO bonding pool in this case). If freezing/pausing of tokens on a network level is a consideration, this should form part of the judgement on whether or not to `allowCollateral` as opposed to increasing complexity (administration and contract size).

Reviewer 3: Internal discussion leads me to believe that it never happened that a major token had frozen trading.
Possibly this is too much detail for this proposal and we can leave sorting out this issue as an implicit part of the execution of the grant.

    // --- SubPool authed functions ---

    /**
     * Indicate a sub-pool's intent to quit the batch auction
     * At this point the solver whose sub-pool it is is now removed from
     * the competition.
     * This then starts the `delay` timer until exit is allowed.
     * @param pool The sub-pool to announce an exit for
     * @dev Emits an event `AnnounceExit`
     */
    function announceDestroy(SubPool pool) external onlySubpoolAuth;
    /**
     * Withdraw any remaining collateral / cow from the sub-pool.
     * Funds are sent to `msg.sender`.
     * @dev Should emit an event `SubPoolDestroyed` on the factory
     */
    function finalizeExit(SubPool pool) external onlySubpoolAuth;
    /**
     * Set the backend URI for the sub-pool
     * @param pool The sub-pool to set the URI for
     * @param newUri The new URI to set
     * @dev Emits an event `SetBackendUri` on the factory when setting a URI
     */
    function setBackendUri(SubPool pool, string calldata newUri) external onlySubpoolAuth;
}

> interface SignedSettlement is Auth {

Reviewer 3: I believe there's no authentication needed?

```suggestion
interface SignedSettlement {
```

     * `SignedSettlement`
     */
    function canSolve(address candidate) external view returns (bool);
    /**
     * View function for determining the sub-pool for a given solver.
     * Reverts if the input address has no sub-pool to its name.
     * @param solver The solver to retrieve the sub-pool for
     * @return The contract managing the collateral paid by this solver to be
     * allowed to settle transactions on CoW Protocol
     */
    function solverSubPool(address solver) external view returns (SubPool);
    /**
     * View function for determining the backend URI for the sub-pool
     * @return string URI
     */
    function backendUri(SubPool pool) external view returns (string memory);

    // --- permissionless functions ---
    /**
     * Deploy a `SubPool` deterministically based on `msg.sender`.
     * @notice The `msg.sender` SHOULD NOT be a contract.
     * @param token The nominated token to use for collateral
     * @param amt How much collateral to add to the sub-pool (the call will revert if
     * this value is below the minimal collateral needed for `token`)
     * @param cowAmt Amount of COW to add as collateral

>      * @param backendUri The URI to send the batches to the solver

Reviewer 3: Do we want to have a pool address determined solely by the solver address? In case, I'd state it somewhere here.

Author: I've relaxed this a little to have the pool address determined with `CREATE` (and not deterministic) to allow for key rotations.

     * @dev Emits an event `SubPoolCreated` when creating a sub-pool
     */
    function create(IERC20 token, uint256 amt, uint256 cowAmt, string calldata backendUri) external;

    // --- cow dao bonding pool authed functions ---

    /**
     * Set the minimum delay before a solver can claim back the collateral from its
     * sub-pool
     * @dev Emits an event when setting a delay
     * @param delay The delay in seconds
     */
    function setExitDelay(uint256 delay) external auth;
    /**
     * Allow `token` to be used as collateral for the sub-pool.
     * @dev Emits an event `AllowCollateral` when allowing/updating a token
     */
    function allowCollateral(IERC20 token, uint256 min) external auth;
    /**
     * Revoke `token` as a collateral for sub-pools
     * @dev Emits an event `RevokeCollateral` when revoking a token
     */

>     function revokeCollateral(IERC20 token) external auth;

Reviewer 3: This means that there should also be functions in the pool that allow pulling out "old" collateral and replacing it with new collateral.

Author: I'm not so sure, I think this is muddied a bit by the 365 day gradual accumulation towards the target deposit bond, in which case then it's not clear as to how to determine the minimum collateral. I'd think in this (presumably rare) case, it would be advisable that they just exit and create a new subpool.

Reviewer 2: We might wanna consider the exit delay for invalidated collateral then and if it would allow a more faster exit.

Author: This is no longer the case as what is, and is not, acceptable collateral will be handled off-chain. Collateralisation (`canSolve`) is only affected by if there is an outstanding bill or not.

# RFP-01: Settlement contract wrapper

## Preamble

Requests for proposals are not to be taken as prescriptive or exhaustive. The community is encouraged to submit proposals that build upon the ideas presented in this document. The scope of the project may change based on the proposal received. The primary intent of this document is to provide a starting point from which to achieve the goals outlined, and what is ultimately implemented may differ from the initial proposal.

## Background

The CoW DAO has recently passed [CIP-44: Reduced bonding requirements](https://snapshot.org/#/cow.eth/proposal/0x1b6f1171633ec3d20c4370db37074aa1bd830486d4d0d6c26165915cc42d9412), which allows solvers to participate in the batch auction with reduced bonding requirements. The reduced bonding requirements are as follows:

- $50k in yield-bearing stable coins or ETH
- 500k COW

The target deposit is:

- $100k in yield-bearing stable coins or ETH
- 1M COW

## Goal

> The goal of this project is to implement a `SubPoolFactory` that allows solvers to participate in the batch auction with reduced bonding requirements. The `SubPoolFactory` will deploy a `SubPool` for each solver, which will be funded to at least the minimum required amount. The `SubPoolFactory` will also provide a wrapper around the `GPv2Settlement` contract that allows solvers to execute their solutions by providing a signature attesting to the solution and their parameters.

Reviewer 3: ```suggestion
The goals of this project are the following.
1. Implement a `SubPoolFactory` that allows solvers to participate in the batch auction with reduced bonding requirements. The `SubPoolFactory` will deploy a `SubPool` for each solver, which will be funded to at least the minimum required amount.
2. Implement `SignedSettlement`, a wrapper around the `GPv2Settlement` contract that allows solvers to execute their solutions by providing a signature attesting to the solution and their parameters.
```

> The goal of this project is to implement a `SubPoolFactory` that allows solvers to participate in the batch auction with reduced bonding requirements. The `SubPoolFactory` will deploy a `SubPool` for each solver, which will be funded to at least the minimum required amount. The `SubPoolFactory` will also provide a wrapper around the `GPv2Settlement` contract that allows solvers to execute their solutions by providing a signature attesting to the solution and their parameters.

Reviewer: ```suggestion
The goal of this project is to implement a `SubPoolFactory` that allows solvers to participate in the batch auction with reduced bonding requirements. The `SubPoolFactory` will deploy a `SubPool` for each solver, which can be funded with arbitrary tokens and amounts. The `SubPoolFactory` will also provide a wrapper around the `GPv2Settlement` contract that allows solvers to execute their solutions by providing a signature attesting to the solution and their parameters.

Verification of meeting the bonding requirements happens off-chain by the component providing that attesting signature.
```

### Key Features

- Ultra gas efficient signed settlement function
- No co-mingling of funds between solvers
- Easily observable on-chain addresses for MEV Blocker rebates

> - Easier access to the batch auction with reduced bonding requirements

Reviewer 3: ```suggestion
- The `SignedSettlement` wrapper gas overhead must be minimal. 
- Solver funds must be only stored in `SubPool`s and not mingled together.
- It must be easy to map solver addresses from and to `SubPool` addresses.  
```

### Deliverables

- Contract system for `SubPoolFactory`, `SubPool`, and `SignedSettlement`
- Comprehensive documentation
- Open-source code repository

### Interface

The interfaces below are a _starting point_ for the project and are not exhaustive / binding - they have not be audited or exhaustively tested / reasoned. The community is encouraged to submit proposals that build upon the ideas presented in this document.

```solidity
interface Auth {
    /**
     * Allows `usr` to perform root-level functions
     */
    function addOwner(address usr) external auth;
    /**
     * Denies `usr` from being able to perform root-level functions
     */
    function removeOwner(address usr) external auth;
    /**
     * Determine if `usr` is able to perform root-level functions
     */
    function isOwner(address usr) external view returns (bool);
}

interface SubPool is Auth {
    /**
     * Determine the amount required to `heal` the pool
     * @return amt of immutable collateralToken
     * @return cowAmt of COW
     */

>     function dues() external view returns (uint256 amt, uint256 cowAmt);

Reviewer 2: without minimum configuration for collateral amounts, this can only return difference in cow amounts from minimum amount.

Reviewer: I think this should return the damage that has been invoiced (which doesn't necessarily need to know the minimal amounts)

    /**
     * Pull in from `msg.sender` the dues required to `heal` the pool
     */
    function heal() external;

> }

Reviewer 3: I'd consider dropping this as it doesn't add anything. We only care about basic fund management and this is now done in the factory.

interface SubPoolFactory is Auth {
    // --- view functions ---
    /**
     * Determine if a candidate:
     * 1. Has a registered sub-pool; and

>      * 2. The sub-pool has not announced an exit; and
>      * 3. The sub-pool has no overdue bills

Reviewer 3: Checking bills is an extra feature, from the discussion in this PR I think this will be checked offchain.

```suggestion
     * 2. The sub-pool has not announced an exit.
```

Reviewer 2: yes, if we need the check onchain, we need to determine how it checks overdue

Author: If they have a `bill` which isn't `heal`'d, they should not be able to solve (this prevents off-chain infra from having to a full view of their "health" and provides a semi-permissionless way for them to remedy the issue, instead of requiring manual intervention.

     * solver and is allowed to submit settlement on CoW Protocol through
     * `SignedSettlement`.
     * @param candidate The address to check
     * @return bool if the address should be allowed to submit settlements through
     * `SignedSettlement`
     */     
    function canSolve(address candidate) external view returns (bool);
    /**
     * View function for determining the sub-pool for a given solver.
     * Reverts if the input address has no sub-pool to its name.
     * @param solver The solver to retrieve the sub-pool for
     * @return The contract managing the collateral paid by this solver to be
     * allowed to settle transactions on CoW Protocol
     */
    function solverSubPool(address solver) external view returns (SubPool);
    /**
     * View function for determining the backend URI for the sub-pool
     * @return string URI
     */
    function backendUri(SubPool pool) external view returns (string memory);

    // --- permissionless functions ---
    /**
     * Deploy a `SubPool` for a solver and fund it with the minimum required collateral.
     * @notice Deposited collateral consists of `amt` of `token` and `cowAmt` of COW.

>      * @param token The nominated yield bearing token to use for collateral

Reviewer 3: Too specific imho, it sounds like we want to enforce that.

```suggestion
     * @param token The nominated token to use for collateral
```

     * @param cowAmt Amount of COW to add as collateral (the call will revert if this
     * value is below the minimal collateral needed for COW - acts as spam prevention
     * and gate-keeps `canSolve`).
     * @param solver The address of the solver to create a sub-pool for (the call will
     * revert if the solver already has a sub-pool or is a contract).
     * @param backendUri The URI to send the batches to the solver
     * @dev Emits an event `SubPoolCreated` when creating a sub-pool.
     */
    function create(
        IERC20 token,
        uint256 amt,
        uint256 cowAmt,
        address solver,
        string calldata backendUri
    ) external;

    // --- cow dao bonding pool authed functions ---

    /**
     * Set the minimum delay before a solver can claim back the collateral from its
     * sub-pool
     * @dev Emits an event when setting a delay
     * @param delay The delay in seconds
     */
    function setExitDelay(uint256 delay) external auth;
    /**
     * Set the minimum collateral required for $COW
     * @dev Emits an event when setting the minimum collateral
     * @param cowAmt The minimum amount of COW required
     */

>     function setMinCow(uint256 cowAmt) external auth;

Reviewer: Is this still needed given we got rid on enforcing concrete bonding amounts?

Author: Having this provides some spam prevention and increases cost for some attacks (notably `canSolve` would be protected by the minimum cow, giving a determined value of which attacks must exceed to be profitable - i.e. if the off-chain infra was compromised / had security flaws)

Reviewer: I don't really see how spam protection plays a role here (I assume the off-chain component tracks a fixed list of addresses it wants to check if they can solve), but in this case I think it should be possible to set min requirements for a bunch of tokens (not just CoW). However, then it becomes a question how `or` relations ships (e.g. 500k COW and 100kUSDC or 100kUSDT) should be modelled, which again adds a lot of complexity, which is why I think we should drop it.

If this is mainly to predict future use of view functions (ie canSolve), how about making this contract inherit from [StorageAccessible](https://github.com/gnosis/util-contracts/blob/main/contracts/storage/StorageAccessible.sol) so we can extend it with view functions later on (and even deploy `canSolve` functions with hardcoded min amounts).

Author: It's simply that it doesn't make much sense to be able to call `canSolve(bob)` and it return `true` if bob has just called `create` but transferred nothing. I'd very much wish to avoid the huge complexity of the if this then or that for different types of collateral.

    /**
     * Bill a sub-pool and send the proceeds to the caller
     * The `auth` is unable to bill the sub-pool if the sub-pool has announced an exit
     * and the `delay` has elapsed.
     * @param pool Which sub-pool to bill
     * @param tokens The tokens to bill
     * @param amts The amount of tokens to bill
     * @param reason The reason for the bill (fine, etc)
     * @dev Emits an event `Bill` when billing a sub-pool
     */

>     function bill(SubPool pool, IERC20[] calldata tokens, uint256[] calldata amts, string calldata reason) external auth;

Reviewer: support for native token?

     * @param tokens The tokens to withdraw
     */
    function exit(SubPool pool, IERC20[] calldata tokens) external onlySubpoolAuth;
    /**
     * Set the backend URI for the sub-pool
     * @param pool The sub-pool to set the URI for
     * @param newUri The new URI to set
     * @dev Emits an event `SetBackendUri` on the factory when setting a URI
     */
    function setBackendUri(SubPool pool, string calldata newUri) external onlySubpoolAuth;
}

interface SignedSettlement {
    /**
     * Signed settlement function that is callable by `msg.sender` if `msg.sender` is in
     * receipt of a valid signature. The signer is a trusted entity and bears the responsibility
     * of ensuring that `msg.sender` is correctly collateralised.
     *  
     * The signed data **MUST** include:
     * - `solver` - the address of the sub-bonded solver, i.e. `msg.sender`
     * - `deadline` - the deadline by which the settlement must be executed (e.g. block number)
     * - `tokens` - the traded tokens vector for the settlement
     * - `clearingPrices` - the clearing prices vector for the settlement
     * - `trades` - the trades vector for the settlement

>      * - `interactions` - the interactions vector for the settlement

Reviewer 3: Do we want to mention the fact that optionally we'd like some of the interactions to be optional or not?

Author: The context for this is intentionally left open, as I expect both options to be analysed with a concrete gas difference so that:
1. There is quantitative analysis on what it would cost to do optional interactions, vs not.
2. This can then be put to backend, in the event that optional interactions is more expensive, for a decision as to whether or not this is fundamentally worth it.

Reviewer 3: There is no context anywhere in this document that says we're interested in only signing some of the interactions (which is almost a business requirement according to [this](https://github.com/cowdao-grants/rfps/pull/1#discussion_r1670008722)).
I'd expect a grantee to just sign what we say needs to be signed. At most, an enterprising grantee would suggest to trim out all interactions and deal with the risks for cheaper execution. But I really don't see why anyone would suggest only signing some: it's clearly much more work and is likely to be significantly more complex or a lot more gas expensive.

interface SubPoolFactory is Auth {
    // --- view functions ---
    /**
     * Determine if a candidate:
     * 1. Has a registered sub-pool; and
     * 2. The sub-pool has not announced an exit; and
     * 3. The sub-pool has no overdue bills
     * While not enforced on-chain, an address for which this call returns `true` is a
     * solver and is allowed to submit settlement on CoW Protocol through
     * `SignedSettlement`.
     * @param candidate The address to check
     * @return bool if the address should be allowed to submit settlements through
     * `SignedSettlement`

>      */     

Reviewer 3: ```suggestion
     */
```
